ktbridge/plugin/src/main/kotlin/app/cash/quickjs/ktbridge/plugin/OutboundCallRewriter.kt

```diff
@@ -0,0 +1,354 @@
+import org.jetbrains.kotlin.backend.common.ir.createImplicitParameterDeclarationWithWrappedDescriptor
+import org.jetbrains.kotlin.descriptors.DescriptorVisibilities
+import org.jetbrains.kotlin.descriptors.Modality
+import org.jetbrains.kotlin.ir.ObsoleteDescriptorBasedAPI
+import org.jetbrains.kotlin.ir.UNDEFINED_OFFSET
+import org.jetbrains.kotlin.ir.builders.IrBlockBodyBuilder
+import org.jetbrains.kotlin.ir.builders.Scope
+import org.jetbrains.kotlin.ir.builders.declarations.addConstructor
+import org.jetbrains.kotlin.ir.builders.declarations.addDispatchReceiver
+import org.jetbrains.kotlin.ir.builders.declarations.addFunction
+import org.jetbrains.kotlin.ir.builders.declarations.addValueParameter
+import org.jetbrains.kotlin.ir.builders.declarations.buildClass
+import org.jetbrains.kotlin.ir.builders.irBlock
+import org.jetbrains.kotlin.ir.builders.irCall
+import org.jetbrains.kotlin.ir.builders.irGet
+import org.jetbrains.kotlin.ir.builders.irInt
+import org.jetbrains.kotlin.ir.builders.irString
+import org.jetbrains.kotlin.ir.builders.irTemporary
+import org.jetbrains.kotlin.ir.declarations.IrClass
+import org.jetbrains.kotlin.ir.declarations.IrDeclarationOrigin
+import org.jetbrains.kotlin.ir.declarations.IrFactory
+import org.jetbrains.kotlin.ir.declarations.IrField
+import org.jetbrains.kotlin.ir.declarations.IrSimpleFunction
+import org.jetbrains.kotlin.ir.expressions.IrCall
+import org.jetbrains.kotlin.ir.expressions.IrExpressionBody
+import org.jetbrains.kotlin.ir.expressions.IrStatementOrigin
+import org.jetbrains.kotlin.ir.types.classFqName
+import org.jetbrains.kotlin.ir.types.defaultType
+import org.jetbrains.kotlin.ir.types.typeWith
+import org.jetbrains.kotlin.ir.util.constructors
+import org.jetbrains.kotlin.ir.util.defaultType
+import org.jetbrains.kotlin.ir.util.functions
+import org.jetbrains.kotlin.ir.util.getPackageFragment
+import org.jetbrains.kotlin.ir.util.primaryConstructor
+import org.jetbrains.kotlin.ir.util.properties
+import org.jetbrains.kotlin.name.FqName
+import org.jetbrains.kotlin.name.Name
+
+/**
+ * Rewrites calls to `createJsService` with two arguments:
+ *
+ * ```
+ * val helloService = createJsClient<EchoService>(EchoJsAdapter, "testing")
+ * ```
+ *
+ * to create a new subtype of JsClient:
+ *
+ * ```
+ * val helloService = object : JsClient<EchoService>(
+ *   jsAdapter = EchoJsAdapter,
+ *   webpackModuleName = "testing",
+ *   packageName = "app.cash.quickjs.ktbridge.testing",
+ *   propertyName = "yoService"
+ * ) {
+ *   override fun get(internalBridge: InternalBridge): EchoService {
+ *     return object : EchoService {
+ *       override fun echo(request: EchoRequest): EchoResponse {
+ *         val outboundCall = OutboundCall(jsAdapter, internalBridge, "echo", 1)
+ *         outboundCall.parameter(request)
+ *         return outboundCall.invoke()
+ *       }
+ *     }
+ *   }
+ * }
+ * ```
+ */
+@ObsoleteDescriptorBasedAPI // TODO(jwilson): is there an alternative?
```

Comment:
Rewrite to owner types like you did the other? At least in a follow-up. Unless there's some other reason for keeping it for now...

Reply:
Yeah, lemme do that right now.

Reply:
Done.